python/lsst/afw/table/schema/schema.cc

```diff
@@ -0,0 +1,522 @@
+            os << self;
+            return os.str();
+        }
+    );
+
+    // Key
+    PyKey<T> clsKey(mod, ("Key" + suffix).c_str());
+    mod.attr("_Key")[pySuffix] = clsKey;
+    clsKey.def(py::init<>());
+    clsKey.def(
+        "__eq__",
+        [](Key<T> const & self, Key<T> const & other)-> bool { return self == other; },
+        py::is_operator()
+    );
+    clsKey.def(
+        "__ne__",
+        [](Key<T> const & self, Key<T> const & other)-> bool { return self != other; },
+        py::is_operator()
+    );
+    clsKey.def("isValid", &Key<T>::isValid);
+    clsKey.def("getOffset", &Key<T>::getOffset);
+    clsKey.def(
+        "__repr__",
+        [](Key<T> const & self) -> std::string {
+            std::ostringstream os;
+            os << self;
+            return os.str();
+        }
+    );
+    // The Key methods below actually wrap templated methods on Schema and
+    // SchemaMapper.  Rather than doing many-type overload resolution by
+    // wrapping those methods directly, we use the visitor pattern by having
+    // the wrappers for those methods delegate back to these non-templated
+    // methods on the templated Key classes.
+    clsKey.def(
+        "_findIn",
+        [](Key<T> const & self, Schema const & schema) {
+            return schema.find(self);
+        }
+    );
+    clsKey.def(
+        "_addMappingTo",
+        [](Key<T> const & self, SchemaMapper & mapper, Field<T> const & field, bool doReplace) {
+            return mapper.addMapping(self, field, doReplace);
+        }
+    );
+    clsKey.def(
+        "_addMappingTo",
+        [](Key<T> const & self, SchemaMapper & mapper, std::string const & name, bool doReplace) {
+            return mapper.addMapping(self, name, doReplace);
+        }
+    );
+    clsKey.def(
+        "_addMappingTo",
+        [](Key<T> const & self, SchemaMapper & mapper, py::object const &, bool doReplace) {
+            return mapper.addMapping(self, doReplace);
+        }
+    );
+    declareKeySpecializations(clsKey);
+
+    // SchemaItem
+    PySchemaItem<T> clsSchemaItem(mod, ("SchemaItem"+suffix).c_str());
+    mod.attr("_SchemaItem")[pySuffix] = clsSchemaItem;
+    clsSchemaItem.def_readonly("key", &SchemaItem<T>::key);
+    clsSchemaItem.def_readonly("field", &SchemaItem<T>::field);
+    clsSchemaItem.def("getKey", [](SchemaItem<T> const & self) { return self.key; });
+    clsSchemaItem.def("getField", [](SchemaItem<T> const & self) { return self.field; });
+    clsSchemaItem.def(
+        "__getitem__",
+        [](py::object const & self, int index) -> py::object {
+            if (index == 0) {
+                return self.attr("key");
+            } else if (index == 1) {
+                return self.attr("field");
+            }
+            // Have to raise IndexError not some LSST exception to get the
+            // right behavior when unpacking.
+            PyErr_SetString(PyExc_IndexError, "Index to SchemaItem must be 0 or 1.");
```

Comment:
It's for element unpacking: `key, field = item`.

I agree it's pretty dangerous (like any unpacking), since you could do `field, key = item` and get something incorrect with no error (yet).  But this was in the old interface and I think we don't want to change it now.